sdk/src/types/block/semantic.rs

```diff
@@ -17,91 +17,110 @@ use crate::types::block::{
+impl std::error::Error for TransactionFailureError {}
 
-/// Represents the different reasons why a transaction can conflict with the ledger state.
+/// Describes the reason of a transaction failure.
 #[repr(u8)]
 #[derive(Debug, Copy, Clone, Eq, PartialEq, packable::Packable)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-#[packable(unpack_error = ConflictError)]
-#[packable(tag_type = u8, with_error = ConflictError::InvalidConflict)]
-pub enum ConflictReason {
-    /// The block has no conflict.
-    None = 0,
-    /// The referenced Utxo was already spent.
+#[packable(unpack_error = TransactionFailureError)]
+#[packable(tag_type = u8, with_error = TransactionFailureError::InvalidReason)]
+pub enum TransactionFailureReason {
+    /// The referenced UTXO was already spent.
     InputUtxoAlreadySpent = 1,
-    /// The referenced Utxo was already spent while confirming this milestone.
-    InputUtxoAlreadySpentInThisMilestone = 2,
-    /// The referenced Utxo cannot be found.
-    InputUtxoNotFound = 3,
+    /// The transaction is conflicting with another transaction. Conflicting specifically means a double spend
+    /// situation that both transaction pass all validation rules, eventually losing one(s) should have this reason.
+    ConflictingWithAnotherTx = 2,
+    /// The referenced UTXO is invalid.
+    InvalidReferencedUtxo = 3,
+    /// The transaction is invalid.
+    InvalidTransaction = 4,
     /// The created amount does not match the consumed amount.
-    CreatedConsumedAmountMismatch = 4,
+    CreatedConsumedAmountMismatch = 5,
     /// The unlock signature is invalid.
-    InvalidSignature = 5,
+    InvalidSignature = 6,
     /// The configured timelock is not yet expired.
-    TimelockNotExpired = 6,
+    TimelockNotExpired = 7,
     /// The given native tokens are invalid.
-    InvalidNativeTokens = 7,
-    /// Storage deposit return mismatch.
-    StorageDepositReturnUnfulfilled = 8,
+    InvalidNativeTokens = 8,
+    /// The return amount in a transaction is not fulfilled by the output side.
+    StorageDepositReturnUnfulfilled = 9,
     /// An invalid unlock was used.
-    InvalidUnlock = 9,
+    InvalidUnlock = 10,
     /// The inputs commitments do not match.
-    InputsCommitmentsMismatch = 10,
+    InputsCommitmentsMismatch = 11,
     /// The sender was not verified.
-    UnverifiedSender = 11,
+    UnverifiedSender = 12,
     /// The chain state transition is invalid.
-    InvalidChainStateTransition = 12,
+    InvalidChainStateTransition = 13,
+    /// The referenced input is created after transaction issuing time.
+    InvalidTransactionIssuingTime = 14,
+    /// The mana amount is invalid.
+    InvalidManaAmount = 15,
+    /// The Block Issuance Credits amount is invalid.
+    InvalidBlockIssuanceCreditsAmount = 16,
+    /// Reward Input is invalid.
+    InvalidRewardInput = 17,
+    /// Commitment Input is invalid.
+    InvalidCommitmentInput = 18,
```

Comment:
I know it's like that in the TIP but I'd personally throw in a `Context` as it's not just an `Input`...

Reply:
Let's change the TIP!

Reply:
Changed and added suggestion to the TIP https://github.com/iotaledger/tips-draft/pull/24/files#r1299181453